### PR TITLE
Add bundled Java runtime

### LINKED ISSUE #468: Bundled Java Runtime
Placeholder to track the progress of bundling the Java Runtime into the installers.

* [x] Download and extract the `target` JDK:<br>**Note:** `ant` makes this tricky.  See [`download-javafx`](https://github.com/qzind/tray/blob/2.1/ant/javafx.xml#L2) as an example.
   * `ant nsis` - Download Windows 64-bit JDK
   * `ant pkgbuild` - Download MacOS 64-bit JDK
   * `ant makeself` - Download Linux 64-bit JDK
* [x] Locate required JRE components using `jlink`, bundle into `<INSTALL_DIR>\jre` folder.
* [x] Ensure all installers use same JRE detection technique (see the NSIS [`FindJRE` function](https://github.com/qzind/tray/blob/2.1/ant/windows/windows-launcher.nsi.in#L55)) (done via #504)
* [x] Installer to use the bundled `java`, `javaw`, `keytool` when available (covered elsewhere)
* [ ] Work through deployment nuances
* [ ] Pass macOS notarization

### PR DESCRIPTION
**Pull request by tresf**
Closes #468 

Bundles the Java Runtime into a the QZ Tray installer.
* To skip bundling, use new parameter`${jre.skip}`
   ```bash
   ant pkgbuild -Djre.skip=true # skip bundling of Java Runtime
   ```
* Since #783 allows JavaFX to be fetched for non-x86_64 architecture by providing new `jre.arch` parameter. e.g:
   ```bash
   ant pkgbuild -Dtarget.arch=aarch64 # Grab and bundle ARM64 Java Runtime + JavaFX
   ```
   See also https://github.com/qzind/tray/commit/b40dd01c64f5b2f2048cc49543c3bf3c6a688995.
* Downloads Java for the architecture requested (or detected) matching the [vendor hard-coded into JLink](https://github.com/qzind/tray/blob/b40dd01c64f5b2f2048cc49543c3bf3c6a688995/src/qz/build/JLink.java#L28-L29).
* Java will now live in a nested folder:
   | OS       | Path |
   |--------|-----|
   | 🪟 Windows | `C:\Program Files\QZ Tray\runtime\bin\java.exe` |
   | 🐧 Linux | `/opt/qz-tray/jre/runtime/java` |
   | 🍏 MacOS | `/Applications/QZ Tray.app/Contents/PlugIns/Java.runtime/Contents/Home/bin/java` |

TODO:

* [x] Update tutorials, don't assume `java -jar` is a valid command.

### CODE REVIEW

File: src/qz/utils/SystemUtilities.java [hunk 1]
```diff
@@ -156,30 +156,22 @@ public static Version getJavaVersion(String version) {
...
     /**
      * Determines the currently running Jar's absolute path on the local filesystem
+     * todo: make this return a sane directory for running via ide
      *
      * @return A String value representing the absolute path to the currently running
      * jar
      */
-    public static String detectJarPath() {
+    public static String getJarPath() {
+        if (jarPath != null) {
```
Comment (tresf) [inline src/qz/utils/SystemUtilities.java:165]: This is redundant, just guard the whole statement, always return the same value.

**Comment by tresf** [comment]
@rebeccadefield as part of this bundled Java runtime effort, @Vzor- and I will be adding an experimental DMG file to test out our sandboxed app with on Apple.  This means we'll need some branding backgrounds, similar to these: [1](https://github.com/LMMS/lmms/blob/master/cmake/apple/dmg_branding.png), [2](https://github.com/LMMS/lmms/blob/master/cmake/apple/dmg_branding%402x.png). 

As in the above examples, the style, placement and size is of our choosing.  Can you provide a 1x and 2x background?   Happy to work offline to prevent too much public back and forth.

**Comment by RebeccaDeField** [comment]
@tresf I'm on it! Will reach out to you to get it started.

**Comment by tresf** [comment]
> @tresf I'm on it! Will reach out to you to get it started.

👍 

Placeholderr artwork is [here](https://raw.githubusercontent.com/qzind/tray/119483cc451512089338543d09086b67c9e6ece7/ant/apple/dmg-background.png) and [here](https://raw.githubusercontent.com/qzind/tray/119483cc451512089338543d09086b67c9e6ece7/ant/apple/dmg-background@2x.png).

File: ant/apple/installer.xml [hunk 2]
```diff
@@ -199,6 +159,45 @@
...
+            <arg value="${build.dir}/appdmg.json"/>
+            <arg value="${out.dir}/${project.filename}${build.type}-${build.version}${codesign.qz.suffix}.dmg"/>
+        </exec>
+    </target>
+
+    <target name="codesign-jre" if="codesign.mac" unless="jre.skip">
+        <property file="ant/apple/apple.properties"/>
+    </target>
+
+    <target name="copy-bundle">
```
Comment (tresf) [inline ant/apple/installer.xml:186]: What was wrong with the old target name `build-app`?  I think it's more descriptive.

Comment (Vzor-) [reply ant/apple/installer.xml:186]: https://github.com/qzind/tray/blob/e2b0f5a4a2aea510b6c502898f40c4aca161e592/ant/apple/installer.xml#L20 it is stiill there, git diff is a bit messy

Comment (tresf) [reply ant/apple/installer.xml:186]: The commit message says `remove redundant code`, yet this block is still redundant.

```xml
        <!-- set payload files executable -->
        <chmod perm="a+x" type="file">
            <fileset dir="${build.copydir}">
                <include name="**/${project.name}"/>
                <include name="**/Resources/uninstall"/>
                <include name="**/bin/*"/>
                <include name="**/lib/jspawnhelper"/>
            </fileset>
        </chmod>
```

Comment (Vzor-) [reply ant/apple/installer.xml:186]: missed that somehow. it is not needed.

**Comment by klabarge** [comment]
I get the below error during installation on Ubuntu 16.0.4

```
QZ Tray is starting...
[INFO] 2021-08-31 19:10:23,401 @ qz.installer.Installer:?
	Fixing runtime permissions...
[INFO] 2021-08-31 19:10:23,407 @ qz.installer.Installer:?
	Stopping running instances...
[DEBUG] 2021-08-31 19:10:23,434 @ qz.utils.ShellUtilities:?
	Executing: [pgrep, -x, java]
[DEBUG] 2021-08-31 19:10:23,461 @ qz.utils.ShellUtilities:?
	Executing: [pgrep, -f, qz-tray.jar]
      [success] Task preinstall was successful
Running install task...
Looking for Java...
   [success] Java was found: /tmp/selfgz7723/runtime/bin/java
Verifying the Java version is 1.8+...
   [success] Java 11.0 was detected
   [success] Java >= 9.0, adding jigsaw options
QZ Tray is starting...
[INFO] 2021-08-31 19:10:24,080 @ qz.installer.Installer:?
	Installing to /opt/qz-tray
[ERROR] 2021-08-31 19:10:24,126 @ qz.utils.ArgParser:?
	Installation step INSTALL failed
java.io.FileNotFoundException: /tmp/selfgz7723/Java.runtime/Contents/Home/legal/java.desktop/ADDITIONAL_LICENSE_INFO (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(Unknown Source)
	at java.base/java.io.FileInputStream.<init>(Unknown Source)
	at org.apache.commons.io.FileUtils.doCopyFile(FileUtils.java:1138)
	at org.apache.commons.io.FileUtils.doCopyDirectory(FileUtils.java:1428)
	at org.apache.commons.io.FileUtils.doCopyDirectory(FileUtils.java:1426)
	at org.apache.commons.io.FileUtils.doCopyDirectory(FileUtils.java:1426)
	at org.apache.commons.io.FileUtils.doCopyDirectory(FileUtils.java:1426)
	at org.apache.commons.io.FileUtils.doCopyDirectory(FileUtils.java:1426)
	at org.apache.commons.io.FileUtils.doCopyDirectory(FileUtils.java:1426)
	at org.apache.commons.io.FileUtils.copyDirectory(FileUtils.java:1389)
	at org.apache.commons.io.FileUtils.copyDirectory(FileUtils.java:1261)
	at org.apache.commons.io.FileUtils.copyDirectory(FileUtils.java:1230)
	at qz.installer.Installer.deployApp(Unknown Source)
	at qz.installer.Installer.install(Unknown Source)
	at qz.installer.Installer.install(Unknown Source)
	at qz.utils.ArgParser.processInstallerArgs(Unknown Source)
	at qz.utils.ArgParser.intercept(Unknown Source)
	at qz.App.main(Unknown Source)
      [failure] Task install failed.
```

**Comment by tresf** [comment]
@klabarge the error you've linked seems to be a very odd issue with `ant`'s inability to remove two files:
* Symbolic links
* `.DS_Store` files

This isn't documented in `ant`'s documentation so it took a while to find, but for now, I've written a workaround to just get rid of these symlinks from the source as well as to always delete `.DS_Store` files: https://github.com/qzind/tray/pull/770/commits/f40bb1e27b156f164d1045e81c6bc1b323e0000d.

This means you'll need to run `ant clean` once, and then the `Java.runtime` folder pollution should stop. 😅 I found a few other edge-case issues with the build system and they've been patched as well.

**Pull request merged**